In [1]:
import tweepy
import numpy as np
import pandas as pd
import requests
import json

In [2]:
df = pd.read_csv('twitter-archive-enhanced.csv')
df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  8.924210e+17                    NaN                  NaN   
1  8.921770e+17                    NaN                  NaN   
2  8.918150e+17                    NaN                  NaN   
3  8.916900e+17                    NaN                  NaN   
4  8.913280e+17                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [160]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1993 entries, 0 to 1993
Data columns (total 19 columns):
tweet_id                      1993 non-null int64
in_reply_to_status_id         23 non-null float64
in_reply_to_user_id           23 non-null float64
timestamp                     1993 non-null datetime64[ns]
source                        1993 non-null object
text                          1993 non-null object
retweeted_status_id           144 non-null float64
retweeted_status_user_id      144 non-null float64
retweeted_status_timestamp    144 non-null datetime64[ns]
expanded_urls                 1993 non-null object
rating_numerator              1993 non-null int64
rating_denominator            1993 non-null int64
name                          1993 non-null object
doggo                         1993 non-null object
floofer                       1993 non-null object
pupper                        1993 non-null object
puppo                         1993 non-null object
fav_count       

In [7]:
df.dtypes

tweet_id                      float64
in_reply_to_status_id         float64
in_reply_to_user_id           float64
timestamp                      object
source                         object
text                           object
retweeted_status_id           float64
retweeted_status_user_id      float64
retweeted_status_timestamp     object
expanded_urls                  object
rating_numerator                int64
rating_denominator              int64
name                           object
doggo                          object
floofer                        object
pupper                         object
puppo                          object
dtype: object

Now we need to download the predictions of each tweet and is present is "image_predictions.tsv", which should be downloaded using requests library.

In [5]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'

In [6]:
r = requests.get(url)

In [7]:
with open('predictions.tsv', mode='wb') as f:
    f.write(r.content)

In [3]:
pred = pd.read_csv('predictions.tsv', sep='\t')
pred.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

So our second dataframe has been gathered. We have 1 dataframe, which is our twitter archive, and another dataframe  which contains image predictions for each tweet, which we have downloaded and saved in a tsv file and then saved in a dataframe.

In [3]:
df.name.value_counts()

None        745
a            55
Charlie      12
Lucy         11
Oliver       11
Cooper       11
Lola         10
Penny        10
Tucker       10
Winston       9
Bo            9
the           8
Sadie         8
Bailey        7
Toby          7
Buddy         7
Daisy         7
an            7
Koda          6
Rusty         6
Jack          6
Dave          6
Milo          6
Jax           6
Oscar         6
Stanley       6
Bella         6
Scout         6
Leo           6
Chester       5
           ... 
Staniel       1
Stark         1
Berb          1
Jo            1
Swagger       1
Kaia          1
Cheryl        1
Godi          1
Eevee         1
Tebow         1
Cilantro      1
Toffee        1
Benny         1
BeBe          1
Aubie         1
Barry         1
Rudy          1
Callie        1
Stefan        1
Grizz         1
Mac           1
Chase         1
Storkson      1
Tayzie        1
Willy         1
Jeremy        1
Loomis        1
General       1
Dunkin        1
Major         1
Name: name, Length: 957,

Many dog names are 'a' in the dataset, which cannot be a dog name.

In [29]:
for i in range(10,20):
    print(df.expanded_urls[i])

https://twitter.com/dog_rates/status/890006608113172480/photo/1,https://twitter.com/dog_rates/status/890006608113172480/photo/1
https://twitter.com/dog_rates/status/889880896479866881/photo/1
https://twitter.com/dog_rates/status/889665388333682689/photo/1
https://twitter.com/dog_rates/status/889638837579907072/photo/1,https://twitter.com/dog_rates/status/889638837579907072/photo/1
https://twitter.com/dog_rates/status/889531135344209921/photo/1
https://twitter.com/dog_rates/status/889278841981685760/video/1
https://twitter.com/dog_rates/status/888917238123831296/photo/1
https://twitter.com/dog_rates/status/888804989199671297/photo/1,https://twitter.com/dog_rates/status/888804989199671297/photo/1
https://twitter.com/dog_rates/status/888554962724278272/photo/1,https://twitter.com/dog_rates/status/888554962724278272/photo/1,https://twitter.com/dog_rates/status/888554962724278272/photo/1,https://twitter.com/dog_rates/status/888554962724278272/photo/1
https://twitter.com/dog_rates/status/887

Quite a few cells have multiple http addresses, and this issue must be resolved. Also, addresses not belonging to weRateDogs page must be removed.

In [4]:
df.tweet_id.sample(5)

2221    6.684800e+17
1449    6.961010e+17
333     8.327570e+17
583     8.001890e+17
1281    7.088110e+17
Name: tweet_id, dtype: float64

In [6]:
print(int(df.tweet_id[0]))   #tweetID of first row in df   
print(df.expanded_urls[0])   #URL address of first row

892421000000000000
https://twitter.com/dog_rates/status/892420643555336193/photo/1


Converting the twitter ID's into int directly leads to loss in precision. So the twitter id's cannot be used as such.

In [7]:
urls_count = df.expanded_urls.apply(lambda x : len(str(x).split(',')))
## this list stores how many urls are stored in each cell under expanded_urls, as multiple url's are seperated by ','. 

In [8]:
urls_count.value_counts()

1    1717
2     365
4     134
3     120
5      10
6       6
8       4
Name: expanded_urls, dtype: int64

In [4]:
df[df.tweet_id.duplicated()].shape

(7, 17)

In [49]:
df.shape

(2356, 17)

In [158]:
w = df.doggo=='doggo'
x = df.pupper == 'pupper'
y = df.puppo == 'puppo'
z = df.floofer == 'floofer'
df[ w | x | y | z].shape

(306, 19)

Now doing assessment of pred dataframe.

In [162]:
pred.sample(5)

tweet_id                                            jpg_url  \
284   671141549288370177    https://pbs.twimg.com/media/CVBfrU9WUAApDeV.jpg   
1635  806219024703037440    https://pbs.twimg.com/media/CzBD7MWVIAA5ptx.jpg   
1308  753655901052166144    https://pbs.twimg.com/media/CnWGCpdWgAAWZTI.jpg   
822   693109034023534592  https://pbs.twimg.com/ext_tw_video_thumb/69310...   
333   672160042234327040    https://pbs.twimg.com/media/CVP9_beUEAAwURR.jpg   

      img_num                  p1   p1_conf  p1_dog                      p2  \
284         1          guinea_pig  0.387728   False             wood_rabbit   
1635        1                chow  0.835102    True              Pomeranian   
1308        1  miniature_pinscher  0.456092    True             toy_terrier   
822         1      cocker_spaniel  0.740013    True  Welsh_springer_spaniel   
333         1                 pug  0.561027    True          French_bulldog   

       p2_conf  p2_dog                  p3   p3_conf  p3_dog  
284   0.171681   False              borzoi  0.075358    True  
1635  0.040783    True          Eskimo_dog  0.021274    True  
1308  0.153126    True   Italian_greyhound  0.144147    True  
822   0.088739    True    golden_retriever  0.047470    True  
333   0.222114    True  Labrador_retriever  0.065456    True

In [179]:
pred.dtypes

tweet_id      int64
jpg_url      object
img_num       int64
p1           object
p1_conf     float64
p1_dog         bool
p2           object
p2_conf     float64
p2_dog         bool
p3           object
p3_conf     float64
p3_dog         bool
dtype: object

In [170]:
size1 = pred[pred.p1_dog==True].shape[0]
pred[pred.p1_dog==True]

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
30    666418789513326592    https://pbs.twimg.com/media/CT-YWb7U8AA7QnN.jpg   
31    666421158376562688    https://pbs.twimg.com/media/CT-aggCXAAIMfT3.jpg   
32    666428276349472768    https://pbs.twimg.com/media/CT-g-0DUwAEQdSn.jpg   
34    666435652385423360    https://pbs.twimg.com/media/CT-nsTQWEAEkyDn.jpg   
35    666437273139982337    https://pbs.twimg.com/media/CT-pKmRWIAAxUWj.jpg   
36    666447344410484738    https://pbs.twimg.com/media/CT-yU5QWwAEjLX5.jpg   
37    666454714377183233    https://pbs.twimg.com/media/CT-5Bs-WUAA2JeC.jpg   
38    666644823164719104    https://pbs.twimg.com/media/CUBl6IwVAAA9_zT.jpg   
...                  ...                                                ...   
2041  885311592912609280    https://pbs.twimg.com/media/C4bTH6nWMAAX_bJ.jpg   
2042  885528943205470208    https://pbs.twimg.com/media/DEoH3yvXgAAzQtS.jpg   
2043  885984800019947520    https://pbs.twimg.com/media/DEumeWWV0AA-Z61.jpg   
2044  886258384151887873    https://pbs.twimg.com/media/DEyfTG4UMAE4aE9.jpg   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsA

In [175]:
size2 = pred[(pred.p1_dog==False) & (pred.p2_dog==True) & (pred.p3_dog==True)].shape[0]
pred[(pred.p1_dog==False) & (pred.p2_dog==True) & (pred.p2_dog==True)]

tweet_id                                            jpg_url  \
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
33    666430724426358785    https://pbs.twimg.com/media/CT-jNYqW4AAPi2M.jpg   
43    666776908487630848    https://pbs.twimg.com/media/CUDeDoWUYAAD-EM.jpg   
52    666996132027977728    https://pbs.twimg.com/media/CUGlb6iUwAITEbW.jpg   
95    667550904950915073    https://pbs.twimg.com/media/CUOb_gUUkAACXdS.jpg   
103   667806454573760512    https://pbs.twimg.com/media/CUSGbXeVAAAgztZ.jpg   
108   667878741721415682    https://pbs.twimg.com/media/CUTILFiWcAE8Rle.jpg   
113   667915453470232577    https://pbs.twimg.com/media/CUTpj-GWcAATc6A.jpg   
133   668480044826800133    https://pbs.twimg.com/media/CUbrDWOWcAEyMdM.jpg   
145   668625577880875008    https://pbs.twimg.com/media/CUdvambWoAA007z.jpg   
150   668641109086707712    https://pbs.twimg.com/media/CUd9ivxWUAAuXSQ.jpg   
163   668960084974809088    https://pbs.twimg.com/media/CUifpn4WUAAS5X3.jpg   
167   668986018524233728    https://pbs.twimg.com/media/CUi3PIrWoAAPvPT.jpg   
171   668994913074286592    https://pbs.twimg.com/media/CUi_UtnWIAEtfqz.jpg   
179   669324657376567296    https://pbs.twimg.com/media/CUnrN7vUcAAfGvN.jpg   
183   669353438988365824    https://pbs.twimg.com/media/CUoFZTyW4AE70iD.jpg   
198   669625907762618368    https://pbs.twimg.com/media/CUr9NjgU8AEpf5w.jpg   
213   670040295598354432    https://pbs.twimg.com/media/CUx2F6lVEAAvFev.jpg   
235   670427002554466305    https://pbs.twimg.com/media/CU3VzVwWwAAAsst.jpg   
237   670433248821026816    https://pbs.twimg.com/media/CU3be0SWEAEqb7I.jpg   
238   670434127938719744    https://pbs.twimg.com/media/CU3cSG8W4AIAePH.jpg   
239   670435821946826752    https://pbs.twimg.com/media/CU3d0azWUAA38FD.jpg   
261   670786190031921152    https://pbs.twimg.com/media/CU8ceuxWUAALMEo.jpg   
263   670792680469889025    https://pbs.twimg.com/media/CU8iYi2WsAEaqQ0.jpg   
270   670822709593571328    https://pbs.twimg.com/media/CU89schWIAIHQmA.jpg   
281   671122204919246848    https://pbs.twimg.com/media/CVBOFTLWwAAzlNi.jpg   
315   671735591348891648    https://pbs.twimg.com/media/CVJ79MzW4AEpTom.jpg   
329   672082170312290304    https://pbs.twimg.com/media/CVO3KodXAAAj1de.jpg   
331   672125275208069120    https://pbs.twimg.com/media/CVPeX2dWwAEwyaR.jpg   
352   672538107540070400    https://pbs.twimg.com/media/CVVV1wJWoAEcOyk.jpg   
...                  ...                                                ...   
1640  807059379405148160    https://pbs.twimg.com/media/Ct2qO5PXEAE6eB0.jpg   
1647  808733504066486276    https://pbs.twimg.com/media/Czky0v9VIAEXRkd.jpg   
1672  813112105746448384    https://pbs.twimg.com/media/C0jBJZVWQAA2_-X.jpg   
1675  813157409116065792    https://pbs.twimg.com/media/C0jqVVOXUAAGJ0G.jpg   
1676  813172488309972993    https://pbs.twimg.com/media/C0j4EESUsAABtMq.jpg   
1699  816829038950027264    https://pbs.twimg.com/media/CvoBPWRWgAA4het.jpg   
1705  817423860136083457  https://pbs.twimg.com/ext_tw_video_thumb/81742...   
1711  818536468981415936    https://pbs.twimg.com/media/C1wGkYoVQAAuC_O.jpg   
1747  823322678127919110    https://pbs.twimg.com/media/C20HmaKWgAQ6-6X.jpg   
1748  823581115634085888    https://pbs.twimg.com/media/C23ypm6VQAAO31l.jpg   
1761  826115272272650244    https://pbs.twimg.com/media/C3bzVILWcAUjS5i.jpg   
1779  828409743546925057    https://pbs.twimg.com/media/C38aQYgXAAMY2Wh.jpg   
1781  828708714936930305    https://pbs.twimg.com/media/C4AqLSgVYAEg8nt.jpg   
1782  828770345708580865    https://pbs.twimg.com/media/C4BiOXOXAAAf6IS.jpg   
1821  834574053763584002    https://pbs.twimg.com/media/C5UAqgyXAAAbMWH.jpg   
1823  834931633769889797    https://pbs.twimg.com/media/C5ZF4p-XEAEmApg.jpg   
1824  835152434251116546    https://pbs.twimg.com/media/C5cOtWVWMAEjO5p.jpg   
1826  835264098648616962    https://pbs.twimg.com/media/C5d0QtvXMAI_7uz.jpg   
1848  839549326359670784    https://pbs.twimg.com/media/C6atpTLWYA

In [177]:
size3 = pred[(pred.p1_dog==False) & (pred.p2_dog==False) & (pred.p3_dog==False)].shape[0]
pred[(pred.p1_dog==False) & (pred.p2_dog==False) & (pred.p2_dog==False)]

tweet_id                                            jpg_url  \
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
45    666786068205871104    https://pbs.twimg.com/media/CUDmZIkWcAAIPPe.jpg   
50    666837028449972224    https://pbs.twimg.com/media/CUEUva1WsAA2jPb.jpg   
51    666983947667116034    https://pbs.twimg.com/media/CUGaXDhW4AY9JUH.jpg   
53    667012601033924608    https://pbs.twimg.com/media/CUG0bC0U8AAw2su.jpg   
56    667065535570550784    https://pbs.twimg.com/media/CUHkkJpXIAA2w3n.jpg   
69    667188689915760640    https://pbs.twimg.com/media/CUJUk2iWUAAVtOv.jpg   
73    667369227918143488    https://pbs.twimg.com/media/CUL4xR9UkAEdlJ6.jpg   
77    667437278097252352    https://pbs.twimg.com/media/CUM2qWaWoAUZ06L.jpg   
78    667443425659232256    https://pbs.twimg.com/media/CUM8QZwW4AAVsBl.jpg   
87    667524857454854144    https://pbs.twimg.com/media/CUOGUfJW4AA_eni.jpg   
93    667549055577362432    https://pbs.twimg.com/media/CUOcVCwWsAERUKY.jpg   
94    667550882905632768    https://pbs.twimg.com/media/CUObvUJVEAAnYPF.jpg   
96    667724302356258817    https://pbs.twimg.com/media/CUQ7tv3W4AA3KlI.jpg   
98    667766675769573376    https://pbs.twimg.com/media/CURiQMnUAAAPT2M.jpg   
100   667782464991965184    https://pbs.twimg.com/media/CURwm3cUkAARcO6.jpg   
106   667866724293877760    https://pbs.twimg.com/media/CUS9PlUWwAANeAD.jpg   
107   667873844930215936    https://pbs.twimg.com/media/CUTDtyGXIAARxus.jpg   
112   667911425562669056    https://pbs.twimg.com/media/CUTl5m1WUAAabZG.jpg   
115   667937095915278337    https://pbs.twimg.com/media/CUT9PuQWwAABQv7.jpg   
117   668142349051129856    https://pbs.twimg.com/media/CUW37BzWsAAlJlN.jpg   
118   668154635664932864    https://pbs.twimg.com/media/CUXDGR2WcAAUQKz.jpg   
123   668226093875376128    https://pbs.twimg.com/media/CUYEFlQXAAUkPGm.jpg   
126   668256321989451776    https://pbs.twimg.com/media/CUYflCXWEAAzQVu.jpg   
...                  ...                                                ...   
1869  844580511645339650    https://pbs.twimg.com/media/C7iNfq1W0AAcbsR.jpg   
1872  844979544864018432    https://pbs.twimg.com/media/C7n4aQ0VAAAohkL.jpg   
1886  847962785489326080    https://pbs.twimg.com/media/C8SRpHNUIAARB3j.jpg   
1887  847971574464610304    https://pbs.twimg.com/media/C8SZH1EWAAAIRRF.jpg   
1891  849051919805034497    https://pbs.twimg.com/media/C8hwNxbXYAAwyVG.jpg   
1892  849336543269576704    https://pbs.twimg.com/media/C8lzFC4XcAAQxB4.jpg   
1899  851224888060895234    https://pbs.twimg.com/media/C9AohFoWsAUmxDs.jpg   
1900  851464819735769094    https://pbs.twimg.com/media/C9ECujZXsAAPCSM.jpg   
1902  851861385021730816    https://pbs.twimg.com/media/C8W6sY_W0AEmttW.jpg   
1905  852226086759018497  https://pbs.twimg.com/ext_tw_video_thumb/85222...   
1906  852311364735569921    https://pbs.twimg.com/media/C9QEqZ7XYAIR7fS.jpg   
1910  853299958564483072    https://pbs.twimg.com/media/C9eHyF7XgAAOxPM.jpg   
1931  859074603037188101    https://pbs.twimg.com/media/C-wLyufW0AA546I.jpg   
1936  860184849394610176    https://pbs.twimg.com/media/C-_9jWWUwAAnwkd.jpg   
1937  860276583193509888    https://pbs.twimg.com/media/C_BQ_NlVwAAgYGD.jpg   
1940  860924035999428608    https://pbs.twimg.com/media/C_KVJjDXsAEUCWn.jpg   
1942  861288531465048066  https://pbs.twimg.com/ext_tw_video_thumb/86128...   
1944  861769973181624320    https://pbs.twimg.com/media/CzG425nWgAAnP7P.jpg   
1946  862457590147678208    https://pbs.twimg.com/media/C_gQmaTUMA

In [178]:
print(size1)# top predicition being dog
print(size2)# top prediction not dog, but 2nd and 3rd prediction being dog 
print(size3)# all predictions not dog

1532
101
324


There are 1532 rows in pred for which the top prediction is a dog, while there are 101 more rows for which top prediction is not a dog, but the 2nd and 3rd high prediction is a dog. Upon visiting the URL's of these pics, it seems that each time the second and third prediction of dog classification true, it <b>is</b> a dog.

### QUALITY

#### ASSESSING

1) Some cells in expanded_url have 2 or more urls, sometimes they are same and sometimes they represent different websites. Need to keep only single correct twitter url.

2) Some expanded_urls are not of twitter. Even of those belonging to twitter, some don't belong to WeRateDogs

3) Tweet id's are in float64, and stored as 8.92421e+17, so cannot access tweet status using it. What can be done is extract the tweet id from the expanded_url column.

4) Many tweet id's are repeated, occur twice in the dataset.

5) Timestamp data is stored in object data type.

6) Retweeted timestamp is stored as object type.

7) Many dog names are 'a', which is not possible. 

8) 324 rows in pred dataframe not corresponding to that of a dog.

#### CLEANING


1) Need to extract correct twitter http using regex.

2) Need to remove rows which don't have twitter http.

3) Need to extract correct tweetID from the twitter http

4) Need to remove duplicate tweet ID's from dataset.

5) Convert timestamp data to datetime format.

6) Convert retweeted timestamp to datetime format.

7) Convert 'a' named dogs to 'None', since those dogs names are not known actually.

8) Rows not that of a dog must be removed.

### TIDINESS

#### ASSESSING

1) The dog type(pupper/puppo) is stored as different columns.

2) img_num in pred dataframe doesn't convey any required information.

#### CLEANING

1) Ideally, condense into single column. However, since very few of them actually have any value, its better to drop the columns of dog type.

2) Drop img_num column.


The accurate URL's for weRateDogs are:

['https://twitter.com/dog_rates/status/711694788429553666/photo/1'] should be accepted

['https://twitter.com/dog_rates/status/752334515931054080/video/1'] should be accepted

Doesn't belong to WeRateDogs:

['https://twitter.com/micahgrimes/status/817902080979599361'] should not be accepted

['https://vine.co/v/iAP0Ugzi2PO'] should not be accepted


In [4]:
df_copy = df.copy()  ##create a copy of the dataset

In [5]:
regex = '(https://twitter.com/dog_rates/status/([0-9]+)/(photo|video)/1)'#regex to extract URL and tweetID

In [6]:
df['expanded_urls'] = df['expanded_urls'].str.extract(regex, expand=True)[0]#first capture group is URL

In [7]:
df['tweet_id'] = df['expanded_urls'].str.extract(regex, expand=True)[1]#second capture group is tweet ID

In [9]:
df[df.expanded_urls.isna()].shape
# number of rows having NaN value in url address

(218, 17)

Need to remove those rows corresponding to non-twitter http addresses. These rows now have NaN value in expanded_urls, as the regex did not match with its http address.

In [8]:
ind = df[df.expanded_urls.isna()].index

In [9]:
df.drop( ind , inplace=True )

In [10]:
df[df.expanded_urls.isna()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [34]:
df.shape

(2138, 17)

In [21]:
df.tweet_id.value_counts()

829374341691346946    2
825026590719483904    2
712809025985978368    2
782969140009107456    2
835264098648616962    2
685325112850124800    2
679158373988876288    2
670319130621435904    2
767754930266464257    2
783839966405230592    2
795076730285391872    2
698195409219559425    2
750429297815552001    2
725842289046749185    2
671896809300709376    2
759923798737051648    2
878281511006478336    2
667509364010450944    2
790946055508652032    2
670444955656130560    2
873213775632977920    2
694669722378485760    2
819004803107983360    2
777684233540206592    2
831650051525054464    2
842163532590374912    2
830583320585068544    2
786963064373534720    2
793286476301799424    2
679828447187857408    2
                     ..
833479644947025920    1
828372645993398273    1
670290420111441920    1
857393404942143489    1
680934982542561280    1
755206590534418437    1
748575535303884801    1
876838120628539392    1
803638050916102144    1
793271401113350145    1
8822681101993697

In [11]:
df[df.duplicated(subset=['tweet_id'])].shape
# tweetID's which have been duplicated in the dataset

(144, 17)

In [12]:
df.drop_duplicates(subset = ['tweet_id'], keep='first', inplace=True)

In [13]:
df[df.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [37]:
df.shape

(1994, 17)

In [14]:
df.timestamp = pd.to_datetime(df.timestamp)
df.retweeted_status_timestamp = pd.to_datetime(df.retweeted_status_timestamp)

In [21]:
df[df.name=='None'].shape

(546, 17)

In [23]:
df[df.name=='a'].shape

(55, 17)

In [43]:
df.name.value_counts()

None         546
a             55
Charlie       11
Oliver        10
Lucy          10
Cooper        10
Tucker         9
Penny          9
Winston        8
Sadie          8
Daisy          7
Lola           7
Toby           7
the            7
Bella          6
an             6
Jax            6
Stanley        6
Bo             6
Koda           6
Scout          5
Chester        5
Buddy          5
Dave           5
Bailey         5
Oscar          5
Milo           5
Rusty          5
Louis          5
Leo            5
            ... 
Tycho          1
Ralph          1
Doobert        1
Rodman         1
Storkson       1
Mac            1
Burt           1
Eevee          1
William        1
Banjo          1
Fletcher       1
Layla          1
Maya           1
Pippin         1
Shikha         1
Stephanus      1
Tino           1
Clarkus        1
Spark          1
Chubbs         1
Tebow          1
Thor           1
Cilantro       1
Toffee         1
Benny          1
BeBe           1
Aubie          1
Kallie        

In [15]:
#Now we'll convert all 'a' names to 'None'.

df['name'] = df['name'].apply(lambda x: 'None' if x=='a' else x)

In [48]:
df.name.value_counts()

None         601
Charlie       11
Lucy          10
Cooper        10
Oliver        10
Penny          9
Tucker         9
Winston        8
Sadie          8
Daisy          7
the            7
Toby           7
Lola           7
Bo             6
Jax            6
Bella          6
Koda           6
Stanley        6
an             6
Chester        5
Buddy          5
Rusty          5
Oscar          5
Milo           5
Louis          5
Leo            5
Dave           5
Bailey         5
Scout          5
George         4
            ... 
Doobert        1
Rodman         1
Smiley         1
Ike            1
Willy          1
Storkson       1
Banjo          1
Chase          1
Fletcher       1
Layla          1
Maya           1
Pippin         1
Shikha         1
Stephanus      1
Tino           1
Clarkus        1
Spark          1
Chubbs         1
Tebow          1
Thor           1
Cilantro       1
Toffee         1
Benny          1
BeBe           1
Aubie          1
Kallie         1
Callie         1
Eevee         

In [28]:
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

In [29]:
api = tweepy.API(auth_handler=auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

In [54]:
type(api.get_status(int(tf), tweet_mode='extended'))

tweepy.models.Status

In [31]:
t = 0

with open('tweet_json.txt', mode='w') as f:
    for tweetid in df['tweet_id']:
        try:
            print("Trying "+str(t)+"....")
            content = api.get_status(tweetid)
            f.write(json.dumps(content._json)+'\n')
            t = t + 1
        except:
            print("Error at "+str(t) + " tweetID is "+str(tweetid))
            t = t+1

Trying 0....
Trying 1....
Trying 2....
Trying 3....
Trying 4....
Trying 5....
Trying 6....
Trying 7....
Trying 8....
Trying 9....
Trying 10....
Trying 11....
Trying 12....
Trying 13....
Trying 14....
Trying 15....
Trying 16....
Trying 17....
Trying 18....
Trying 19....
Trying 20....
Trying 21....
Trying 22....
Trying 23....
Trying 24....
Trying 25....
Trying 26....
Trying 27....
Trying 28....
Trying 29....
Trying 30....
Trying 31....
Trying 32....
Trying 33....
Trying 34....
Trying 35....
Trying 36....
Trying 37....
Trying 38....
Trying 39....
Trying 40....
Trying 41....
Trying 42....
Trying 43....
Trying 44....
Trying 45....
Trying 46....
Trying 47....
Trying 48....
Trying 49....
Trying 50....
Trying 51....
Trying 52....
Trying 53....
Trying 54....
Trying 55....
Trying 56....
Trying 57....
Trying 58....
Trying 59....
Trying 60....
Trying 61....
Trying 62....
Trying 63....
Trying 64....
Trying 65....
Trying 66....
Trying 67....
Trying 68....
Trying 69....
Trying 70....
Trying 71....
Tr

In [32]:
nd = [754011816964026368,728760639972315136,669682095984410625]#error at these indexes, so trying again......

In [33]:
with open('tweet_json.txt', mode='a') as f:
    for tweetid in nd:
        try:
            print("Trying "+str(t)+"....")
            content = api.get_status(tweetid)
            f.write(json.dumps(content._json)+'\n')
            t = t + 1
        except:
            print("Error at "+str(t) + " tweetID is "+str(tweetid))
            t = t+1

Trying 1994....
Error at 1994 tweetID is 754011816964026368
Trying 1995....
Trying 1996....


The tweet corresponding to tweetID 754011816964026368 has been deleted, and hence cannot be retrieved. Since the tweet has been deleted, it makes sense to remove it from the dataframe.


In [27]:
index = df[df.tweet_id == 754011816964026368].index
df.drop( index, inplace=True )

In [30]:
df[df.tweet_id==754011816964026368]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo, fav_count, retweet_count]
Index: []


Now we have finished downloading all tweet data for each tweet_id in df using tweepy and stored in tweet_json.txt . Now we have to read that data and extract tweetID, retweet count and favourite count for all of the tweets and store in a dataframe.

In [17]:
tweet_dict = []
with open('tweet_json.txt', mode='r') as f:
    for tweet_data in f:
        json_obj = json.loads(tweet_data)
        tweetID = json_obj['id']
        retweet_count = json_obj['retweet_count']
        fav_count = json_obj['favorite_count']
        
        tweet_dict.append({'tweet_id' : tweetID,
                           'retweet_count' : retweet_count,
                           'fav_count' : fav_count})

In [18]:
df2 = pd.DataFrame(tweet_dict)

In [22]:
type(df2)

pandas.core.frame.DataFrame

In [60]:
df2.head()

fav_count  retweet_count            tweet_id
0      38674           8555  892420643555336193
1      33136           6288  892177421306343426
2      24948           4171  891815181378084864
3      42054           8676  891689557279858688
4      40214           9444  891327558926688256

In [52]:
df2.shape

(1993, 3)

In [26]:
df.shape

(1994, 19)

In [31]:
df_copy.shape

(1993, 19)

In [53]:
df2.dtypes

fav_count        int64
retweet_count    int64
tweet_id         int64
dtype: object

In [80]:
pred.sample(3)

tweet_id                                          jpg_url  \
51    666983947667116034  https://pbs.twimg.com/media/CUGaXDhW4AY9JUH.jpg   
1975  870063196459192321  https://pbs.twimg.com/media/DBMV3NnXUAAm0Pp.jpg   
1027  710844581445812225  https://pbs.twimg.com/media/Cd1tYGmXIAAoW5b.jpg   

      img_num          p1   p1_conf  p1_dog         p2   p2_conf  p2_dog  \
51          1        swab  0.589446   False  chain_saw  0.190142   False   
1975        1  comic_book  0.534409   False   envelope  0.280722   False   
1027        1       dingo  0.536593   False   Pembroke  0.200407    True   

               p3   p3_conf  p3_dog  
51            wig  0.034510   False  
1975  book_jacket  0.043786   False  
1027      basenji  0.060734    True

In [81]:
pred.dtypes

tweet_id      int64
jpg_url      object
img_num       int64
p1           object
p1_conf     float64
p1_dog         bool
p2           object
p2_conf     float64
p2_dog         bool
p3           object
p3_conf     float64
p3_dog         bool
dtype: object

In [28]:
df_copy = df.copy()

In [29]:
df.tweet_id = pd.to_numeric(df.tweet_id)

Now we shall merge the 2 dataframes df and df2 using the twitter ID's of the former.

In [33]:
df = df.merge( df2, on='tweet_id', how='left')
df.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1478  676864501615042560                    NaN                  NaN   
456   801958328846974976                    NaN                  NaN   
113   866686824827068416                    NaN                  NaN   
1991  666033412701032449                    NaN                  NaN   
1900  667782464991965184                    NaN                  NaN   

               timestamp                                             source  \
1478 2015-12-15 20:40:47  <a href="http://twitter.com/download/iphone" r...   
456  2016-11-25 01:18:59  <a href="http://twitter.com/download/iphone" r...   
113  2017-05-22 16:06:55  <a href="http://twitter.com/download/iphone" r...   
1991 2015-11-15 23:21:54  <a href="http://twitter.com/download/iphone" r...   
1900 2015-11-20 19:12:01  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1478  Meet Ash. He's just a head now. Lost his body ...                  NaN   
456   This is Ronnie. He hopes you're having a great...                  NaN   
113   This is Lili. She can't believe you betrayed h...                  NaN   
1991  Here is a very happy pup. Big fan of well-main...                  NaN   
1900  Super rare dog. Endangered (?). Thinks it's fu...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1478                       NaN                        NaT   
456                        NaN                        NaT   
113                        NaN                        NaT   
1991                       NaN                        NaT   
1900                       NaN                        NaT   

                                          expanded_urls  rating_numerator  \
1478  https://twitter.com/dog_rates/status/676864501...                10   
456   https://twitter.com/dog_rates/status/801958328...                12   
113   https://twitter.com/dog_rates/status/866686824...                12   
1991  https://twitter.com/dog_rates/status/666033412...                 9   
1900  https://twitter.com/dog_rates/status/667782464...                 9   

      rating_denominator    name doggo floofer pupper puppo  fav_count  \
1478                  10     Ash  None    None   None  None     2223.0   
456                   10  Ronnie  None    None   None  None     8421.0   
113                   10    Lili  None    None   None  None    19590.0   
1991                  10    None  None    None   None  None      125.0   
1900                  10    None  None    None   None  None      422.0   

      retweet_count  
1478          779.0  
456          1918.0  
113          3586.0  
1991           44.0  
1900          253.0

In [34]:
df.dtypes

tweet_id                               int64
in_reply_to_status_id                float64
in_reply_to_user_id                  float64
timestamp                     datetime64[ns]
source                                object
text                                  object
retweeted_status_id                  float64
retweeted_status_user_id             float64
retweeted_status_timestamp    datetime64[ns]
expanded_urls                         object
rating_numerator                       int64
rating_denominator                     int64
name                                  object
doggo                                 object
floofer                               object
pupper                                object
puppo                                 object
fav_count                            float64
retweet_count                        float64
dtype: object

### ASSESS

1) The fav_count and retweet_count columns have been converted into float after merging with df. 

### CLEAN

1) Convert it into int type.

In [35]:
df.fav_count = df.fav_count.astype(int)
df.retweet_count = df.retweet_count.astype(int)

In [36]:
df.dtypes

tweet_id                               int64
in_reply_to_status_id                float64
in_reply_to_user_id                  float64
timestamp                     datetime64[ns]
source                                object
text                                  object
retweeted_status_id                  float64
retweeted_status_user_id             float64
retweeted_status_timestamp    datetime64[ns]
expanded_urls                         object
rating_numerator                       int64
rating_denominator                     int64
name                                  object
doggo                                 object
floofer                               object
pupper                                object
puppo                                 object
fav_count                              int32
retweet_count                          int32
dtype: object

In [35]:
df.shape

(1993, 21)